function_calling/tests/fc_test.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python3
+"""
+Function Calling (FC) Test Script
+
+This script tests the functionality of the Function Calling using unittest.
+
+Usage:
+    python tests/fc_test.py
+
+Returns:
+    0 if all tests pass, or a positive integer representing the number of failed tests.
+"""
+
+import logging
+import os
+import sys
+from contextlib import contextmanager, redirect_stdout
+from io import StringIO
+#from time import sleep
+import time
+from typing import Callable, Generator, Optional
+import unittest
+
+# Setup logging
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+logger = logging.getLogger(__name__)
+
+# Setup paths and global variables
+current_dir = os.getcwd()
+kit_dir = current_dir # absolute path for function_calling kit dir
+repo_dir = os.path.abspath(os.path.join(kit_dir, '..')) # absolute path for ai-starter-kit root repo
+
+sys.path.append(kit_dir)
+sys.path.append(repo_dir)
+
+from function_calling.src.function_calling import FunctionCallingLlm  # type: ignore
+from function_calling.src.tools import calculator, get_time, python_repl, query_db # type: ignore
+
+tools = [get_time, calculator, python_repl, query_db]
+
+class FCTestCase(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.time_start = time.time()
+        cls.fc = FunctionCallingLlm(tools)
+    
+    # Add assertions
+    def test_fc_initialization(self):
+        self.assertIsInstance(self.fc, FunctionCallingLlm, "fc is not of type FunctionCallingLlm")
+
+    def test_get_time_tool(self):
+        result = get_time.invoke({'kind': 'both'})
+        logger.info("\nInvoking get_time:")
+        logger.info(result)
+        self.assertTrue(result,"get_time result shouldn't be empty")
+    
+    def test_calculator_tool(self):
+        result = calculator.invoke('5*23.7 -5')
+        logger.info("\nInvoking calculator:")
+        logger.info(result)
+        self.assertTrue(result,"calculator result shouldn't be empty")
+
+    def test_python_repl_tool(self):
+        result = python_repl.invoke({'command': 'for i in range(0,5):\n\tprint(i)'})
+        logger.info("\nInvoking python_repl:")
+        logger.info(result)
+        self.assertTrue(result,"python_repl result shouldn't be empty")
+
+    def test_query_db_tool(self):
+        result = query_db.invoke({'query': 'How many genres of music are in the chinook db'})
+        logger.info("\nInvoking query_db:")
+        logger.info(result)
+        self.assertTrue(result,"query_db result shouldn't be empty")
+
+    def test_fc_pipeline(self):
+        response = self.fc.function_call_llm('what time is it?', max_it=5, debug=False)
+        logger.info("\nFunction calling pipeline with get_time:")
+        logger.info(response)
+        self.assertTrue(response,"LLM final response shouldn't be empty")
+
+        response = self.fc.function_call_llm('it is time to go to sleep, how many hours last to 10pm?', max_it=5, debug=False)
```

Comment:
I suggest changing this query to avoid the use of get time in an intermediate step, having slightly different inputs potentially leading the model to have different paths to accomplish the task, without us knowing if this will be accomplished in less than 5 iterations

Reply:
I modified this query to only use the calculator: 
        response = self.fc.function_call_llm('calculate 5*10-3', max_it=5, debug=False)